components/layout_thread/lib.rs

```diff
@@ -1664,35 +1668,46 @@ impl ProfilerMetadataFactory for LayoutThread {
     }
 }
 
-// The default computed value for background-color is transparent (see
-// http://dev.w3.org/csswg/css-backgrounds/#background-color). However, we
-// need to propagate the background color from the root HTML/Body
-// element (http://dev.w3.org/csswg/css-backgrounds/#special-backgrounds) if
-// it is non-transparent. The phrase in the spec "If the canvas background
-// is not opaque, what shows through is UA-dependent." is handled by rust-layers
-// clearing the frame buffer to white. This ensures that setting a background
-// color on an iframe element, while the iframe content itself has a default
-// transparent background color is handled correctly.
-fn get_root_flow_background_color(flow: &mut Flow) -> webrender_api::ColorF {
-    let transparent = webrender_api::ColorF { r: 0.0, g: 0.0, b: 0.0, a: 0.0 };
-    if !flow.is_block_like() {
-        return transparent;
+/// Propagate background if necessary from the <body> to the document root.
+///
+/// Backgrounds by default are transparent. This ensures that iframes
+/// are displayed with a background from the surrounding context.
+///
+/// Spec: https://drafts.csswg.org/css-backgrounds/#body-background
+fn determine_root_background(root_flow: &mut Flow) {
+    if !root_flow.is_block_like() {
+        return;
+    }
+    let root_block = root_flow.as_mut_block();
+    let root_style = ServoArc::make_mut(&mut root_block.fragment.style);
+    {
+        let root_bg = root_style.get_background();
+        // If the root <html> element already has a background do nothing.
+        // The background will be rendered correctly without changes.
+        if root_bg.background_image != get_initial_background_image() ||
+            root_bg.background_color != Color::transparent() {
+            debug!("Some background found for root element.");
+            return;
+        }
     }
+    debug!("Root element without background. Trying to propagate child element background.");
 
-    let block_flow = flow.as_mut_block();
-    let kid = match block_flow.base.children.iter_mut().next() {
-        None => return transparent,
+    let kid = match root_block.base.children.iter_mut().next() {
+        None => return,
+        Some(ref kid) if !kid.is_block_like() => return,
         Some(kid) => kid,
     };
-    if !kid.is_block_like() {
-        return transparent;
-    }
-
-    let kid_block_flow = kid.as_block();
-    kid_block_flow.fragment
-                  .style
-                  .resolve_color(kid_block_flow.fragment.style.get_background().background_color)
-                  .to_gfx_color()
+    let kid_mut_block = kid.as_mut_block();
+    let kid_background = ServoArc::make_mut(&mut kid_mut_block.fragment.style).mutate_background();
```

Comment:
I'm pretty sure this `make_mut` is not needed, is it?

Why wouldn't just `kid_mut_block.fragment.style.get_background()` suffice?

But more importantly:

 * What guarantees that the `<body>` is a block?
 * What guarantees that it's the first block?

Not that the code before wasn't wrong, but...

Reply:
> I'm pretty sure this make_mut is not needed, is it?
> Why wouldn't just kid_mut_block.fragment.style.get_background() suffice?

Well the borrow checker disagrees. In lines 1709 and 1710 `kid_background` is mutated. This prompts the compiler to tell me ``cannot borrow immutable borrowed content `*kid_background` as mutable``.

> What guarantees that the <body> is a block?

Good catch. While the `<body>` is usually a block with `body { display: inline; }` it is not. Which other [`FlowClass`](https://doc.servo.org/layout/flow/enum.FlowClass.html) besides block and inline can it be? 

> What guarantees that it's the first block?

The body element is always the first block. (Even if there is no `<body>` tag in the source code. Or the code is something like `<a href="https://mozilla.org">Mozilla</a><body>Main content</body>`.)
There must be some specifications that describes this. One part is [Optional Tags](https://html.spec.whatwg.org/multipage/syntax.html#optional-tags). Another is [HTML parsing The "after head" insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#the-after-head-insertion-mode) and [HTML parsing The "in body" insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody). Servo seems to follow all these weird rules as far as I can tell.

Reply:
> The body element is always the first block. (Even if there is no <body> tag in the source code. Or the code is something like <a href="https://mozilla.org">Mozilla</a><body>Main content</body>.)

I'm pretty sure I can do `head { display: block }`.

Reply:
Also, with script I can do all sorts of stuff.

Reply:
> I'm pretty sure I can do head { display: block }.

Yes this breaks my assumption "always first block". And afaik I can't even ask flows if they are a body element.